README.md

# http2-cache.js

[![Build Status](https://travis-ci.org/kaazing/http2-cache.js.svg?branch=develop)](http://travis-ci.org/kaazing/http2-cache.js)

[![npm version](https://img.shields.io/npm/v/http2-cache.svg?style=flat)](https://www.npmjs.com/package/http2-cache)

Exposes http caching to the browser by adding functionality to XMLHttpRequest, and then running XMLHttpRequest over http2 over WebSockets

### Warning

#### This library is pre 1.0.0, much of the functionality is not yet fully implemented.

This library provides a user-space based HTTP/2 client implementation and cache that slots
in under traditional network Browser APIs such as XHR (eventually Fetch).  This project is
intended to be useful for 1. providing HTTP/2 emulation when not natively available and 
2. allowing backends to pre-populate a browser-side cache via HTTP/2 push.

This project is designed to be 100% spec compliant.  It should be removable without loss 
of functionality when there is universal support for HTTP/2 with a consistent H2-cache 
implementation.

### Design

HTTP/2 (RFC-7540) runs in userspace by running it on top of WebSocket.  An HTTP (RFC-7234) in
memory cache implementation stores cacheable responses.  The XHR API can be configured to route
a subset of requests via this transport stack.

A consistent implementation for caching HTTP/2 push requests has not yet emerged.  See
[discussion](https://docs.google.com/document/d/1v3rjj0DMDTocUtZSjOwdwt8D-yhCw6R5SVaax4MPgMc/edit)
for current the state of the world. This HTTP/2 implementation will not use the cache for
any request (Including HTTP/2 pushed) that contains the 
[request cache-directive "no-cache"](https://tools.ietf.org/html/rfc7234#section-5.2.1.4). 
HTTP/2 pushed requests that do not include this directive may have their HTTP/2 stream aborted
by the client if the cache already contains a cached response for that request.  I.E. use the
request cache-directive "no-cache" when doing cache-busting.  This should work in all cases where
you want to do a cache replacement.

HTTP/2 push requests require an established stream to send the push request.  The API provides a
means to open a long-lived upstream request to an arbitrary location that may be used to send
push requests.  Alternatively, streams may be left open for sending future pushed responses via
"long-pushing", that is sending the push promise for a future response, prior to completing the
response to an existing request.  I.e. always maintain one response in flight, by sending the push
promise for it prior to completing a response.

### API/Usage

The API attaches to the XMLHttpRequest object.  

```javascript
XMLHttpRequest.proxy([urls of configurations])
```

The `proxy([urls of configurations])` triggers fetching of JSON configurations on the backend
server.  The configurations should be of the following form:

```
{
    // Logger debugLevel true='info' or (info|debug|trace)
    "clientLogLevel": false,
    // Transport endpoint
    "transport": "wss://where-the-underlying-ws-transport-connects:443/",
    // Transport push path
    "push": "optional-path-that-is-opened-for-pushes",
    // Transport reconnect settings
    "reconnect": true,
    "reconnectInterval": 100,
    "maximumReconnectInterval": 4000,
    // AccelerationStrategy default to "always" can be "connected"
    // - Value "always" means always/don't make requests if they are proxied but no ws connection is open. 
    // - Value "connected" means make requests when connected via websocket.
    "accelerationStrategy": "always",
    "proxy": [
      "http://origin-to-send-via-http2:80/path/",
      "http://origin-to-send-via-http2:80/path2/",
      "http://other-origin-to-send-via-http2:80"
    ]
}
```

In full

```html
<script type="text/javascript" src="http2-cache.js"></script>
<script type="text/javascript">
    XMLHttpRequest.proxy(["http://localhost:8000/config"]);
</script>
```

### Build

The integration tests require Java JDK 8 be installed.

```bash
npm i
npm run build
```

### Integration Tests

```bash
npm run test:browser
```


HISTORY.md
Version history
===============

### 0.2.23 (2018-07-15) 
* Better error handling to avoid block http2 streams from flowing
### 0.2.18 (2018-12-11) 
* Use http2.js^4.0.6 from npm to fix SETTINGS_MAX_CONCURRENT_STREAMS

### 0.2.16 (2018-12-11) 
* Use http2.js^4.0.5 from npm to fix SETTINGS_MAX_CONCURRENT_STREAMS

### 0.2.15 (2018-12-06)
* Fix possible cache hit on POST|PUT|DELETE (via #118)

### 0.2.14 (2018-12-06)
* Fix bad merge on revalidate Cache

### 0.2.13 (2018-12-05)

* Cancel duplicate push promises streams #111 (via #113)
* Force new authorization value through cache #112 (via #114).
* Handle Etag to allow server sending NotModified 304 responses #116 (via #115)

### 0.2.12 (2018-07-18) Maintenance release
* Use http2.js^4.0.4 from npm instead of kaazing/http2.js#v4.0.3

### 0.2.11 (2018-07-18) 

* Update to http2.js 4.0.3 to add support on request for `retry-after` header on `503|429|302` status code.

### 0.2.10

* Fix FormData 'multipart/form-data' boundary support.
  Note: FormData limited support, only string supported

### 0.2.8

* Implement AccelerationStrategy default to "always" can be "connected"
AccelerationStrategy value "always" means always/don't make requests if they are proxied but no ws connection is open and value "connected" means make requests when connected via websocket.

### 0.2.7

* We non longer commit index.js generated file

### 0.2.3

* We non run postinstall to build index.js


lib/utils.js
var url = require('url'),
    util = require("util"),
    InvalidStateError = require('./errors').InvalidStateError;


var resolvePort = function (u) {
    u = (u instanceof url.constructor) ? u : url.parse(u);
    var port = u.port;
    if (port === null) {
        if (u.protocol === "ws:" || u.protocol === "http:") {
            port = 80;
        } else {
            port = 443;
        }
    }
    return port;
};
/* global console */
var parseUrl = function (href) {
    var uri = (href instanceof url.constructor) ? href : url.parse(href);
    uri.port = resolvePort(uri);


    if (
        uri.hostname === null &&
            typeof window !== 'undefined'
    ) {
        uri.protocol = window.location.protocol;
        uri.hostname = window.location.hostname;
        uri.port = window.location.port;
        uri.host = uri.hostname + ':' + uri.port;
        uri.href = uri.protocol + "//" + uri.host + uri.href;
    }

    // Define uri.origin
    uri.origin = uri.hostname + ":" + uri.port;
 
    // Check if host match origin (example.com vs example.com:80)
    if (uri.host !== uri.origin) {
        // Fix href to include default port
        uri.href = uri.href.replace(uri.protocol + "//" + uri.host, uri.protocol + "//" + uri.origin);
        // Fix host to include default port
        uri.host = uri.hostname + ":" + uri.port;
    }

    return uri;
};

var redefine = function (obj, prop, value) {

    if (obj[prop]) {
        // TODO, consider erasing scope/hiding (enumerable: false)
        obj["_" + prop] = obj[prop];
    }

    Object.defineProperty(obj, prop, {
        value: value,
        enumerable: obj.propertyIsEnumerable(prop),
        configurable: true
    });
};

var defineGetter = function (obj, prop, getter) {

    if (obj[prop]) {
        // TODO, consider erasing scope/hiding (enumerable: false)
        obj["_" + prop] = obj[prop];
    }

    Object.defineProperty(obj, prop, {
        enumerable: true,
        configurable: true,
        get: getter
    });
};

var definePrivate = function (obj, prop, value) {
    Object.defineProperty(obj, prop, {
        value: value,
        enumerable: false,
        configurable: true
    });
};

var definePublic = function (obj, prop, value) {
    Object.defineProperty(obj, prop, {
        value: value,
        enumerable: true,
        configurable: true
    });
};

var getOrigin = function (u) {
    u = (u instanceof url.constructor) ? u : parseUrl(u);
    return u.protocol + '//' + u.hostname + ':' + resolvePort(u);
};

var defaultPort = function (u, port) {
    var parse = (u instanceof url.constructor) ? u : parseUrl(u);
    if (!port) {
        port = resolvePort(u);
    }
    u = (u instanceof url.constructor) ? u : parseUrl(u);
    return u.protocol + '//' + u.hostname + ':' + port + parse.path;
};

// Import String.fromCodePoint polyfill
require('string.fromcodepoint');

var Utf8ArrayToStr = function (array) {
    var c, char2, char3, char4,
        out = "",
        len = array.length,
        i = 0;
        
    while (i < len) {
        c = array[i++];
        switch (c >> 4) {
            case 0:
            case 1:
            case 2:
            case 3:
            case 4:
            case 5:
            case 6:
            case 7:
                // 0xxxxxxx
                out += String.fromCharCode(c);
                break;
            case 12:
            case 13:
                // 110x xxxx   10xx xxxx
                char2 = array[i++];
                out += String.fromCharCode(((c & 0x1F) << 6) | (char2 & 0x3F));
                break;
            case 14:
                // 1110 xxxx  10xx xxxx  10xx xxxx
                char2 = array[i++];
                char3 = array[i++];
                out += String.fromCharCode(((c & 0x0F) << 12) |
                    ((char2 & 0x3F) << 6) |
                    ((char3 & 0x3F) << 0));
                break;
            case 15:
            // 1111 0xxx 10xx xxxx 10xx xxxx 10xx xxxx
            char2 = array[i++];
            char3 = array[i++];
            char4 = array[i++];
            out += String.fromCodePoint(((c & 0x07) << 18) | ((char2 & 0x3F) << 12) | ((char3 & 0x3F) << 6) | (char4 & 0x3F));
                break;
        }
    }

    return out;
};


var mergeTypedArrays = function (a, b) {
    // Checks for truthy values on both arrays
    if(!a && !b) {
        throw 'Please specify valid arguments for parameters a and b.';  
    }

    // Checks for truthy values or empty arrays on each argument
    // to avoid the unnecessary construction of a new array and
    // the type comparison
    if(!b || b.length === 0) {
        return a;
    }

    if(!a || a.length === 0) {
        return b;
    }

    // Make sure that both typed arrays are of the same type
    if(Object.prototype.toString.call(a) !== Object.prototype.toString.call(b)) {
        throw 'The types of the two arguments passed for parameters a and b do not match.';
    }

    var c = new a.constructor(a.length + b.length);

    // On NodeJS < v4 TypeArray.set is not working as expected, 
    // starting NodeJS 6+ TypeArray.set.length is 1 and works properly.
    if (c.set.length > 0) {
        c.set(a, 0);
        c.set(b, a.length);
    // TypedArray.set relly on deprecated Buffer.set on NodeJS < 6 and producing bad merge
    // Using forEach and Native Setter instead  
    } else {
        a.forEach(function (byte, index) { c[index] = byte; });
        b.forEach(function (byte, index) { c[a.length + index] = byte; });
    }
    
    return c;
};

var toArrayBuffer = function (buf) {
    var ab = new ArrayBuffer(buf.length),
        view = new Uint8Array(ab);

    for (var i = 0; i < buf.length; ++i) {
        view[i] = buf[i];
    }

    return ab;
};

var dataToType = function (data, type) {

    //if payload is missing don't throw runtime exception that will block h2 connection stream execution.
    if (data === undefined) {
        console.error("Reieved response data is undefied");
        return undefined;
    }
    // https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/responseType
    switch (type) {
        case "":
        case "text":
            if (data instanceof Uint8Array) {
                return Utf8ArrayToStr(data);
            }
            return data;
        case "json":
            return JSON.parse(data);
        case "arraybuffer":
            return toArrayBuffer(data);
        case "blob":
            if (typeof Blob !== 'undefined') {
                return new Blob(data);
            } else {
                console.error("Unsupported Response Type: " + type);
                return undefined;
            }
            break;
        case "document":
            if (
                typeof document !== 'undefined' &&
                    typeof document.implementation !== 'undefined' &&
                        typeof document.implementation.createDocument === 'function'
            ) {
                return document.implementation.createDocument('http://www.w3.org/1999/xhtml', 'html', data);
            } else {
                console.error("Unsupported Response Type: " + type);
                return undefined;
            }
            break;
        default:
            console.error("Unexpected Response Type: " + type);
            return undefined;
    }
};

var caseInsensitiveEquals = function (str1, str2) {
    return str1.toUpperCase() === str2.toUpperCase();
};

var serializeXhrBody = function (headers, body) {

    if (typeof body === 'object') {
        
        // Display the key/value pairs
        var bodyParts = [],
            pairEntry, partPair;

        // TODO detect FormData vs POST 
        // FormData serialization
        // application/x-www-form-urlencoded and multipart/form-data.
        // https://gist.github.com/joyrexus/524c7e811e4abf9afe56
        if (typeof body.entries === 'function') {
        
            var key, value,
                iterator = body.entries(),
                boundary = "----webkitformboundary";

            // Use body._boundary if provided for unit-test and custom value 
            boundary += body._TestBoundary || (+(new Date())).toString(16);

            while (
                (pairEntry = iterator.next()) &&
                    pairEntry.done === false
            ) {
                value = pairEntry.value[1]; 
                key = pairEntry.value[0];

                if (typeof value !== "string") {
                    throw new Error("FormData limited support, only string supported (" + typeof value + ')');
                }

                var field = util.format("\r\n--%s\r\n", boundary);
                field += util.format("Content-Disposition: form-data; name=\"%s\"\r\n\r\n", key);
                field += value;
                bodyParts.push(field);
            }

            bodyParts.push(util.format("\r\n--%s--", boundary));
            body = bodyParts.join('');  

            // Refused to set unsafe header "Content-Length"
            //headers.set('Content-Length', body.length);
            headers.set('Content-Type', 'multipart/form-data; boundary=' + boundary);

        } else {

            Object.keys(body).forEach(function (key) {
                var value = body[key];
                if (typeof value  !== "string") {
                    throw new Error("FormData limited support, only string supported (" + typeof value + ')');
                }
                bodyParts.push(encodeURIComponent(key) + '=' + encodeURIComponent(value));
            });

            body = bodyParts.join('&');   
        }
        
        //headers.set('Content-Type', 'application/x-www-form-urlencoded');
    }

    return body;
};

module.exports = {
    parseUrl: parseUrl,
    redefine: redefine,
    defineGetter: defineGetter,
    definePrivate: definePrivate,
    definePublic: definePublic,
    resolvePort: resolvePort,
    getOrigin: getOrigin,
    dataToType: dataToType,
    defaultPort: defaultPort,
    serializeXhrBody: serializeXhrBody,
    caseInsensitiveEquals: caseInsensitiveEquals,
    mergeTypedArrays: mergeTypedArrays,
    Utf8ArrayToStr: Utf8ArrayToStr
};
